### PR TITLE
Benchmarked CLAD gradient tests

### PR DESCRIPTION
**Pull request by 540KJ**
@oshadura Is this what you wanted?

### CODE REVIEW

**Comment by oshadura** [comment]
@540KJ thanks a lot for a benchmark! Can you please register it in CMakeLists.txt file: https://github.com/root-project/rootbench/blob/master/root/hist/hist/CMakeLists.txt, otherwise it is not compiling :)

File: root/hist/hist/TFormulaGradientBenchmarks.cxx [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    auto h = new TF1("f1", "gaus");
+    double p[] = {3, 1, 2};
+    h->SetParameters(p);
+    double x[] = {0};
+    TFormula::GradientStorage result_clad(3);
+    TFormula *f = h->GetFormula();
+    state.ResumeTiming();
+
+    f->GenerateGradientPar();
+    for (unsigned i = 0; i<10000; ++i)
```
Comment (vgvassilev) [inline root/hist/hist/TFormulaGradientBenchmarks.cxx]: You do not need this loop here, we gBenchmark already makes the loop for us.

File: root/hist/hist/TFormulaGradientBenchmarks.cxx [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+static void BM_TFormula_GausTFormulaPerf_Clad(benchmark::State &state) {
+  for (auto _ : state) {
+    state.PauseTiming();
+    auto h = new TF1("f1", "gaus");
+    double p[] = {3, 1, 2};
+    h->SetParameters(p);
+    double x[] = {0};
+    TFormula::GradientStorage result_clad(3);
+    TFormula *f = h->GetFormula();
+    state.ResumeTiming();
```
Comment (vgvassilev) [inline root/hist/hist/TFormulaGradientBenchmarks.cxx]: Could you move all this initialization out of the loop. Our hot function we want to test is `GradientPar`.

File: root/hist/hist/TFormulaGradientBenchmarks.cxx [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+static void BM_TFormula_GausTFormulaPerf_Numerical(benchmark::State &state) {
+  for (auto _ : state) {
+    state.PauseTiming();
+    auto h = new TF1("f1", "gaus");
+    double p[] = {3, 1, 2};
+    h->SetParameters(p);
+    double x[] = {0};
+    TFormula::GradientStorage result_num(3);
+    state.ResumeTiming();
```
Comment (vgvassilev) [inline root/hist/hist/TFormulaGradientBenchmarks.cxx]: Likewise.

File: root/hist/hist/TFormulaGradientBenchmarks.cxx [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+    for (unsigned i = 0; i<10000; ++i)
```
Comment (vgvassilev) [inline root/hist/hist/TFormulaGradientBenchmarks.cxx]: Likewise.

File: root/hist/hist/TFormulaGradientBenchmarks.cxx [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  for (auto _ : state) {
+    state.PauseTiming();
+    auto f1 = new TF1("f1", "gaus");
+    auto h1 = new TH1D("h1", "h1", 1000000, -5, 5);
+    double p1[] = {1, 0, 1.5};
+    f1->SetParameters(p1);
+    h1->FillRandom("f1", 100000);
+    double p2[] = {100, 1, 3};
+    f1->SetParameters(p2);
+    state.ResumeTiming();
```
Comment (vgvassilev) [inline root/hist/hist/TFormulaGradientBenchmarks.cxx]: Likewise.

File: root/hist/hist/TFormulaGradientBenchmarks.cxx [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
```
Comment (vgvassilev) [inline root/hist/hist/TFormulaGradientBenchmarks.cxx]: Likewise.

File: root/hist/hist/TFormulaGradientBenchmarks.cxx [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+    f1->GetFormula()->GenerateGradientPar(); // makes it om par
+    auto r2 = h1->Fit(f1, "S G"); // clad
+
+    state.PauseTiming();
```
Comment (vgvassilev) [inline root/hist/hist/TFormulaGradientBenchmarks.cxx]: Please move that out of the loop.

File: root/hist/hist/TFormulaGradientBenchmarks.cxx [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+  f1->SetParameters(p1);
+  h1->FillRandom("f1", 100000);
+  double p2[] = {100, 1, 3};
+  f1->SetParameters(p2);
+  auto r1 = h1->Fit(f1, "S"); // numerical
+  f1->GetFormula()->GenerateGradientPar(); // makes it om par
+  
+  for (auto _ : state) {
+    auto r2 = h1->Fit(f1, "S G"); // clad
+    ASSERT_FLOAT_EQ(r1->MinFcnValue(), r2->MinFcnValue());
```
Comment (540KJ) [inline root/hist/hist/TFormulaGradientBenchmarks.cxx]: Is putting ASSERT_FLOAT_EQ here fine? Putting it outside loop means r2 is out of scope.

**Comment by 540KJ** [comment]
@oshadura @vgvassilev Added changes.

**Comment by oshadura** [comment]
@540KJ Thank you, much more better! Travis CI is "red" because:
` 54%] Building CXX object root/hist/hist/CMakeFiles/HistBenchmarks.dir/TFormulaGradientBenchmarks.cxx.o
cd /travis/build/root/hist/hist && /usr/bin/g++-7    -I/travis/rootbench/include -I/cvmfs/sft-nightlies.cern.ch/lcg/nightlies/rootcov/Thu/ROOT/HEAD-dfc25/x86_64-ubuntu1604-gcc54-opt/include -I/travis/build/root/hist/hist -I/travis/build/googlebenchmark-prefix/src/googlebenchmark/include  -std=c++14 -pipe -fsigned-char -pthread -fprofile-arcs -ftest-coverage -g    -std=c++14 -pipe -fsigned-char -pthread -fprofile-arcs -ftest-coverage -o CMakeFiles/HistBenchmarks.dir/TFormulaGradientBenchmarks.cxx.o -c /travis/rootbench/root/hist/hist/TFormulaGradientBenchmarks.cxx
/travis/rootbench/root/hist/hist/TFormulaGradientBenchmarks.cxx:6:10: fatal error: gtest/gtest.h: No such file or directory
 #include "gtest/gtest.h"
          ^~~~~~~~~~~~~~~
compilation terminated.
.`

We don't have gtest as a dependency in rootbench.git, can you please remove google test includes and adjust code accordingly please?

**Comment by 540KJ** [comment]
@oshadura Ah, sorry, forgot to comment that I updated. Anyways, was this what you wanted?

**Comment by oshadura** [comment]
@540KJ yes it was, thank you!

**Comment by oshadura** [comment]
@vgvassilev:

4: BM_TFormula_GausTFormulaPerf_Clad              92 ns         92 ns    8226610
4: BM_TFormula_GausTFormulaPerf_Numerical        908 ns        907 ns     784874
4: Info in <TCanvas::MakeDefCanvas>:  created default TCanvas with name c1
4:  FCN=4362.66 FROM MIGRAD    STATUS=CONVERGED     227 CALLS         228 TOTAL
4:                      EDM=4.30011e-08    STRATEGY= 1      ERROR MATRIX ACCURATE 
4:   EXT PARAMETER                                   STEP         FIRST   
4:   NO.   NAME      VALUE            ERROR          SIZE      DERIVATIVE 
4:    1  Constant     1.06152e+00   4.26596e-03   1.10716e-04  -3.99295e-03
4:    2  Mean        -4.79304e-02   1.92496e-01   6.20160e-03   1.69973e-05
4:    3  Sigma        9.47074e+00   8.92402e-01   3.23153e-03   2.27088e-03
4:  FCN=4362.66 FROM MIGRAD    STATUS=CONVERGED      90 CALLS          91 TOTAL
4:                      EDM=2.23954e-08    STRATEGY= 1      ERROR MATRIX ACCURATE 
4:   EXT PARAMETER                                   STEP         FIRST   
4:   NO.   NAME      VALUE            ERROR          SIZE      DERIVATIVE 
4:    1  Constant     1.06152e+00   4.59936e-03   3.54291e-03  -3.08328e-02
4:    2  Mean        -4.79384e-02   1.92515e-01   1.92496e-02  -4.63828e-04
4:    3  Sigma        9.47067e+00   9.42246e-01   1.03303e-01  -4.79540e-04
4: BM_TFormula_GausPerf_Clad              2784215451 ns 2783464559 ns          1
4: Warning in <TROOT::Append>: Replacing existing TH1: h1 (Potential memory leak).
4:  FCN=4328.94 FROM MIGRAD    STATUS=CONVERGED     318 CALLS         319 TOTAL
4:                      EDM=8.55254e-08    STRATEGY= 1      ERROR MATRIX ACCURATE 
4:   EXT PARAMETER                                   STEP         FIRST   
4:   NO.   NAME      VALUE            ERROR          SIZE      DERIVATIVE 
4:    1  Constant     1.06093e+00   4.28107e-03   1.10207e-04  -7.18605e-02
4:    2  Mean        -1.77036e-02   1.98578e-01   6.38100e-03  -4.11277e-04
4:    3  Sigma        9.60654e+00   9.41397e-01   3.41935e-03  -1.03712e-03
4: BM_TFormula_GausPerf_Numerical         2298097503 ns 2298057663 ns          1
1/1 Test #4: rootbench-HistBenchmarks .........   Passed   12.28 sec

**Pull request merged**